File: backend/routers/meta_accounts.py
"""Meta account selection and campaign listing endpoints."""
import json
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete

from database import get_db
from models.auth import AuthToken, ProviderEnum
from models.campaign import Project, AdSet, Ad, ProjectStatus
from schemas.auth import AccountSelectionRequest, MetaCampaignSummary
from services.meta_api import MetaAPIService
from routers.auth import get_meta_token, get_meta_auth

router = APIRouter(prefix="/api/meta", tags=["meta-accounts"])


@router.get("/businesses")
async def list_businesses(db: AsyncSession = Depends(get_db)):
    """List all Business Portfolios accessible with the current token."""
    token = await get_meta_token(db)
    meta = MetaAPIService(token)
    businesses = await meta.get_businesses()
    return {"businesses": businesses}


@router.get("/businesses/{business_id}/ad-accounts")
async def list_business_ad_accounts(
    business_id: str, db: AsyncSession = Depends(get_db)
):
    """List ad accounts for a specific Business Portfolio."""
    token = await get_meta_token(db)
    meta = MetaAPIService(token)
    accounts = await meta.get_business_ad_accounts(business_id)
    return {"ad_accounts": accounts}


@router.get("/pages")
async def list_pages(db: AsyncSession = Depends(get_db)):
    """List Facebook Pages accessible with the current token."""
    token = await get_meta_token(db)
    meta = MetaAPIService(token)
    pages = await meta.get_pages()
    return {"pages": pages}


@router.post("/select-account")
async def select_account(
    data: AccountSelectionRequest, db: AsyncSession = Depends(get_db)
):
    """Save the selected business, ad account, and page."""
    result = await db.execute(
        select(AuthToken).where(AuthToken.provider == ProviderEnum.meta)
    )
    token = result.scalar_one_or_none()
    if not token:
        raise HTTPException(401, "Meta not authenticated")

    token.business_id = data.business_id
    token.business_name = data.business_name
    token.ad_account_id = data.ad_account_id
    if data.page_id:
        token.page_id = data.page_id

    await db.commit()
    return {
        "status": "selected",
        "business_id": data.business_id,
        "business_name": data.business_name,
        "ad_account_id": data.ad_account_id,
        "page_id": token.page_id,
    }


@router.get("/campaigns/{ad_account_id}")
async def list_campaigns(ad_account_id: str, db: AsyncSession = Depends(get_db)):
    """List all campaigns (active + inactive) for an ad account."""
    token = await get_meta_token(db)
    meta = MetaAPIService(token)
    campaigns = await meta.get_campaigns(ad_account_id)
    return {"campaigns": campaigns}


@router.get("/campaign-structure/{campaign_id}")
async def get_campaign_structure(
    campaign_id: str, db: AsyncSession = Depends(get_db)
):
    """Fetch full campaign structure (campaign + ad sets + ads) for Modify mode."""
    token = await get_meta_token(db)
    meta = MetaAPIService(token)
    structure = await meta.get_campaign_structure(campaign_id)
    return structure


@router.post("/import-campaign/{campaign_id}")
async def import_campaign_for_editing(
    campaign_id: str,
    ad_account_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Import an existing Meta campaign into a local Project for editing."""
    auth = await get_meta_auth(db)
    meta = MetaAPIService(auth.access_token, page_id=auth.page_id or "")
    structure = await meta.get_campaign_structure(campaign_id)

    campaign_data = structure.get("campaign", {})

    # Create local Project
    project = Project(
        name=campaign_data.get("name", "Imported Campaign"),
        status=ProjectStatus.previewed,
        campaign_objective=campaign_data.get("objective"),
        daily_budget=float(campaign_data["daily_budget"]) / 100 if campaign_data.get("daily_budget") else None,
        lifetime_budget=float(campaign_data["lifetime_budget"]) / 100 if campaign_data.get("lifetime_budget") else None,
        ad_account_id=ad_account_id,
        business_id=auth.business_id,
        mode="modify",
        meta_source_campaign_id=campaign_id,
        meta_campaign_id=campaign_id,
    )
    db.add(project)
    await db.flush()

    # Create local AdSets and Ads
    for adset_data in structure.get("ad_sets", []):
        targeting = adset_data.get("targeting", {})
        adset = AdSet(
            project_id=project.id,
            name=adset_data.get("name", ""),
            generated_name=adset_data.get("name", ""),
            targeting_json=json.dumps(targeting) if isinstance(targeting, dict) else None,
            optimization_goal=adset_data.get("optimization_goal", "OFFSITE_CONVERSIONS"),
            bid_strategy=adset_data.get("bid_strategy", "LOWEST_COST_WITHOUT_CAP"),
            budget=float(adset_data["daily_budget"]) / 100 if adset_data.get("daily_budget") else None,
            meta_adset_id=adset_data.get("id"),
            status=adset_data.get("status", "PAUSED").lower(),
        )
        db.add(adset)
        await db.flush()

        for ad_data in adset_data.get("ads", []):
            ad = Ad(
                ad_set_id=adset.id,
                name=ad_data.get("name", ""),
                generated_name=ad_data.get("name", ""),
                meta_ad_id=ad_data.get("id"),
                status=ad_data.get("status", "PAUSED").lower(),
            )
            db.add(ad)

    await db.commit()
    await db.refresh(project)

    return {
        "project_id": project.id,
        "name": project.name,
        "mode": "modify",
        "ad_sets_count": len(structure.get("ad_sets", [])),
    }

File: api/lib/routes/meta-accounts.ts
import { Router, Request, Response } from "express";
import { eq } from "drizzle-orm";
import { db } from "../db/client.js";
import { authTokens, projects, adSets, ads } from "../db/schema.js";
import { getMetaToken, getMetaAuth } from "./auth.js";

const router = Router();

// ── GET /businesses ───────────────────────────────────────────────────────

router.get("/businesses", async (_req: Request, res: Response) => {
  try {
    const token = await getMetaToken(db);

    const resp = await fetch(
      `https://graph.facebook.com/v21.0/me/businesses?fields=id,name&access_token=${encodeURIComponent(token)}`,
    );
    const data = (await resp.json()) as any;
    if (data.error) {
      return res.status(400).json({ detail: data.error.message ?? "Failed to fetch businesses" });
    }

    return res.json({ businesses: data.data ?? [] });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── GET /businesses/:id/ad-accounts ───────────────────────────────────────

router.get("/businesses/:id/ad-accounts", async (req: Request, res: Response) => {
  try {
    const businessId = req.params.id;
    const token = await getMetaToken(db);

    const resp = await fetch(
      `https://graph.facebook.com/v21.0/${businessId}/owned_ad_accounts?fields=id,name,account_status,currency&access_token=${encodeURIComponent(token)}`,
    );
    const data = (await resp.json()) as any;
    if (data.error) {
      return res.status(400).json({ detail: data.error.message ?? "Failed to fetch ad accounts" });
    }

    return res.json({ ad_accounts: data.data ?? [] });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── GET /pages ────────────────────────────────────────────────────────────

router.get("/pages", async (_req: Request, res: Response) => {
  try {
    const token = await getMetaToken(db);

    const resp = await fetch(
      `https://graph.facebook.com/v21.0/me/accounts?fields=id,name,access_token&access_token=${encodeURIComponent(token)}`,
    );
    const data = (await resp.json()) as any;
    if (data.error) {
      return res.status(400).json({ detail: data.error.message ?? "Failed to fetch pages" });
    }

    return res.json({ pages: data.data ?? [] });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /select-account ─────────────────────────────────────────────────

router.post("/select-account", async (req: Request, res: Response) => {
  try {
    const { business_id, business_name, ad_account_id, page_id } = req.body;

    if (!business_id || !ad_account_id) {
      return res.status(400).json({ detail: "business_id and ad_account_id are required" });
    }

    const rows = await db
      .select()
      .from(authTokens)
      .where(eq(authTokens.provider, "meta"))
      .limit(1);
    const token = rows[0];
    if (!token) {
      return res.status(401).json({ detail: "Meta not authenticated" });
    }

    const updates: Record<string, unknown> = {
      businessId: business_id,
      businessName: business_name ?? null,
      adAccountId: ad_account_id,
    };
    if (page_id) {
      updates.pageId = page_id;
    }

    await db
      .update(authTokens)
      .set(updates)
      .where(eq(authTokens.id, token.id));

    return res.json({
      status: "selected",
      business_id,
      business_name: business_name ?? null,
      ad_account_id,
      page_id: page_id ?? token.pageId,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── GET /campaigns/:id ────────────────────────────────────────────────────

router.get("/campaigns/:id", async (req: Request, res: Response) => {
  try {
    const adAccountId = req.params.id;
    const token = await getMetaToken(db);

    const resp = await fetch(
      `https://graph.facebook.com/v21.0/${adAccountId}/campaigns?fields=id,name,status,objective,daily_budget,lifetime_budget&access_token=${encodeURIComponent(token)}`,
    );
    const data = (await resp.json()) as any;
    if (data.error) {
      return res.status(400).json({ detail: data.error.message ?? "Failed to fetch campaigns" });
    }

    return res.json({ campaigns: data.data ?? [] });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── GET /campaign-structure/:id ───────────────────────────────────────────

router.get("/campaign-structure/:id", async (req: Request, res: Response) => {
  try {
    const campaignId = req.params.id;
    const token = await getMetaToken(db);

    // Fetch campaign
    const campaignResp = await fetch(
      `https://graph.facebook.com/v21.0/${campaignId}?fields=id,name,status,objective,daily_budget,lifetime_budget&access_token=${encodeURIComponent(token)}`,
    );
    const campaign = (await campaignResp.json()) as any;
    if (campaign.error) {
      return res.status(400).json({ detail: campaign.error.message ?? "Failed to fetch campaign" });
    }

    // Fetch ad sets
    const adsetsResp = await fetch(
      `https://graph.facebook.com/v21.0/${campaignId}/adsets?fields=id,name,status,targeting,optimization_goal,bid_strategy,daily_budget,lifetime_budget&access_token=${encodeURIComponent(token)}`,
    );
    const adsetsData = (await adsetsResp.json()) as any;

    // Fetch ads for each ad set
    const adSetsList = [];
    for (const adset of adsetsData.data ?? []) {
      const adsResp = await fetch(
        `https://graph.facebook.com/v21.0/${adset.id}/ads?fields=id,name,status,creative&access_token=${encodeURIComponent(token)}`,
      );
      const adsData = (await adsResp.json()) as any;

      adSetsList.push({
        ...adset,
        ads: adsData.data ?? [],
      });
    }

    return res.json({
      campaign,
      ad_sets: adSetsList,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /import-campaign/:id ─────────────────────────────────────────────

router.post("/import-campaign/:id", async (req: Request, res: Response) => {
  try {
    const campaignId = req.params.id;
    const { ad_account_id } = req.body;

    if (!ad_account_id) {
      return res.status(400).json({ detail: "ad_account_id is required" });
    }

    const auth = await getMetaAuth(db);
    const token = auth.accessToken;

    // Fetch campaign structure from Meta
    const campaignResp = await fetch(
      `https://graph.facebook.com/v21.0/${campaignId}?fields=id,name,status,objective,daily_budget,lifetime_budget&access_token=${encodeURIComponent(token)}`,
    );
    const campaignData = (await campaignResp.json()) as any;
    if (campaignData.error) {
      return res.status(400).json({ detail: campaignData.error.message ?? "Failed to fetch campaign" });
    }

    // Fetch ad sets with ads
    const adsetsResp = await fetch(
      `https://graph.facebook.com/v21.0/${campaignId}/adsets?fields=id,name,status,targeting,optimization_goal,bid_strategy,daily_budget,lifetime_budget&access_token=${encodeURIComponent(token)}`,
    );
    const adsetsData = (await adsetsResp.json()) as any;

    // Create local Project
    const projectResult = await db
      .insert(projects)
      .values({
        name: campaignData.name ?? "Imported Campaign",
        status: "previewed",
        campaignObjective: campaignData.objective ?? null,
        dailyBudget: campaignData.daily_budget
          ? parseFloat(campaignData.daily_budget) / 100
          : null,
        lifetimeBudget: campaignData.lifetime_budget
          ? parseFloat(campaignData.lifetime_budget) / 100
          : null,
        adAccountId: ad_account_id,
        businessId: auth.businessId ?? null,
        mode: "modify",
        metaSourceCampaignId: campaignId,
        metaCampaignId: campaignId,
      })
      .returning();
    const project = projectResult[0];

    let adSetsCount = 0;

    // Create local AdSets and Ads
    for (const adsetData of adsetsData.data ?? []) {
      const targeting = adsetData.targeting ?? {};

      const adsetResult = await db
        .insert(adSets)
        .values({
          projectId: project.id,
          name: adsetData.name ?? "",
          generatedName: adsetData.name ?? "",
          targetingJson: typeof targeting === "object" ? JSON.stringify(targeting) : null,
          optimizationGoal: adsetData.optimization_goal ?? "OFFSITE_CONVERSIONS",
          bidStrategy: adsetData.bid_strategy ?? "LOWEST_COST_WITHOUT_CAP",
          budget: adsetData.daily_budget
            ? parseFloat(adsetData.daily_budget) / 100
            : null,
          metaAdsetId: adsetData.id ?? null,
          status: (adsetData.status ?? "PAUSED").toLowerCase(),
        })
        .returning();
      const adset = adsetResult[0];
      adSetsCount++;

      // Fetch ads for this ad set
      const adsResp = await fetch(
        `https://graph.facebook.com/v21.0/${adsetData.id}/ads?fields=id,name,status&access_token=${encodeURIComponent(token)}`,
      );
      const adsData = (await adsResp.json()) as any;

      for (const adData of adsData.data ?? []) {
        await db.insert(ads).values({
          adSetId: adset.id,
          name: adData.name ?? "",
          generatedName: adData.name ?? "",
          metaAdId: adData.id ?? null,
          status: (adData.status ?? "PAUSED").toLowerCase(),
        });
      }
    }

    return res.json({
      project_id: project.id,
      name: project.name,
      mode: "modify",
      ad_sets_count: adSetsCount,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

export default router;
